src/main.rs

```diff
@@ -298,10 +308,167 @@ async fn create_endpoint(
     if let Some(addr) = common.ipv6_addr {
         builder = builder.bind_addr_v6(addr);
     }
+    if common.persist || common.persist_at.is_some() {
+        builder = builder.secret_key(get_secret_key(common.persist_at.as_ref()).await);
+    }
     let endpoint = builder.bind().await?;
     Ok(endpoint)
 }
 
+async fn get_secret_key(persist_at: Option<&PathBuf>) -> SecretKey {
+    let persist_at_cow = persist_at
+        .map(Cow::from) // Reference
+        .or_else(|| {
+            std::env::home_dir().map(|mut p| {
+                p.push(".auth");
+                p.push("dumbpipe.key");
+                debug!("Persisting key at: {p:?}");
+                Cow::from(p) // Owned
+            })
+        });
+    let persist_at = persist_at_cow.as_ref().map(Cow::as_ref);
+
+    match read_key(persist_at).await {
+        Ok(Some(result)) => return result,
+        Ok(None) => {}
+        Err(error) => {
+            error!("Error reading persisted dumbpipe key: [{:?}]", error);
+        }
+    }
+
+    let key = SecretKey::generate(&mut rand::rng());
+    if let Some(node_path) = persist_at {
+        if let Err(error) = write_key(node_path, &key).await {
+            error!("Could not persist dumbpipe key: {node_path:?}: {error:?}");
+        }
+    }
+    key
+}
+
+#[derive(Debug, Snafu)]
+#[non_exhaustive]
+pub enum PersistError {
```

Comment:
Because dumbpipe is an application, I think you can remove this error enum and instead just return `n0_snafu::Error`